### ISSUE TITLE
Fix Style/RedundantLineContinuation false positives with pattern matching/string concatenation

### ISSUE THREAD

**Issue by bkuhlmann** [rubocop/rubocop#11756]
## Expected behavior

Hello. :wave: With the new `Style/RedundantLineContinuation` cop enabled, I'm getting a few false positives for line continuation when using pattern matching and/or string concatenation.

## Actual behavior

Consider the following:

``` ruby
module Demo
  def self.demo = "demo"
end

Demo.method(:demo).parameters in [[:rest]] \
                               | [[:keyrest]] \
                               | [[:rest], [:keyrest]] \
                               | [[:rest, :*]] \
                               | [[:keyrest, :**]] \
                               | [[:rest, :*], [:keyrest, :**]]
```

If the line continuation is removed, this will result in the following syntax error:

``` 
syntax error, unexpected '|' (SyntaxError)
```

## Steps to reproduce the problem

Use the following to see the syntax error:

``` ruby
module Demo
  def self.demo = "demo"
end

Demo.method(:demo).parameters in [[:rest]]
                               | [[:keyrest]]
                               | [[:rest], [:keyrest]]
                               | [[:rest, :*]]
                               | [[:keyrest, :**]]
                               | [[:rest, :*], [:keyrest, :**]]
```

If it helps, I'm seeing this in a couple other situations as well. Here's a few more examples:

``` ruby
# RSpec: In this case, removing line continuation would hit a line length issue so definitely want this to be wrapped.
expect(temp_dir.join("test/Procfile.dev").read).to eq(
  %(web: rerun --dir app,config,lib,slices --pattern="**/*.{css,erb,js,rb}" )
  "-- bundle exec puma --config ./config/puma.rb\n"
)
```

## RuboCop version

```
1.49.0 (using Parser 3.2.2.0, rubocop-ast 1.28.0, running on ruby 3.2.2) [arm64-darwin22.4.0]
  - rubocop-capybara 2.17.1
  - rubocop-performance 1.16.0
  - rubocop-rake 0.6.0
  - rubocop-rspec 2.19.0
  - rubocop-thread_safety 0.5.1
```
